### PR TITLE
add to warning banner quota warnings

### PR DESCRIPTION
**Pull request by timothyclemansinsea**
for #1376

### CODE REVIEW

File: src/smc-webapp/project_page.cjsx [hunk 1]
```diff
@@ -187,16 +190,43 @@ FreeProjectWarning = rclass ({name}) ->
         </div>
 
     render : ->
+        if @props.project
+            status       = JSON.parse(JSON.stringify(@props.project.get('status'))) # without parse and stringify not getting this data in the right format
+            total_quotas = @props.get_total_project_quotas(@props.project_id)
+            memory     = '?'
+            disk       = '?'
+            console.log('keys', Object.keys(status));
```
Comment (williamstein) [inline src/smc-webapp/project_page.cjsx]: get rid of all console.log's -- should almost never have this in production.

File: src/smc-webapp/project_page.cjsx [hunk 2]
```diff
@@ -187,16 +190,43 @@ FreeProjectWarning = rclass ({name}) ->
         </div>
 
     render : ->
+        if @props.project
```
Comment (williamstein) [inline src/smc-webapp/project_page.cjsx]: Refresh your browser on a project -- this `@props.project` isn't defined until you move to another page and move back, so the warning banner doesn't appear.

**Comment by timothyclemansinsea** [comment]
@williamstein any idea why @props.project_map isn't defined and how to fix it?

File: src/smc-webapp/project_page.cjsx [hunk 3]
```diff
@@ -186,17 +184,44 @@ FreeProjectWarning = rclass ({name}) ->
             </ul>
         </div>
 
+    render_learn_more : ->
+        <a onClick={=>@actions(project_id: @props.project_id).show_extra_free_warning()}> learn more...</a>
+
     render : ->
+        if @props.project
+            status       = JSON.parse(JSON.stringify(@props.project.get('status'))) # without parse and stringify not getting this data in the right format
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: This looks entirely suspect.

File: src/smc-webapp/project_page.cjsx [hunk 4]
```diff
@@ -186,17 +184,44 @@ FreeProjectWarning = rclass ({name}) ->
...
+                rss = status.memory?.rss
+                if rss
+                    memory = Math.round(rss/1000)
+                disk = status.disk_MB
+                if disk
+                    disk = Math.ceil(disk)
         if not require('./customize').commercial
             return null
         if @props.free_warning_closed
             return null
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: These should be sooner.

File: src/smc-webapp/project_page.cjsx [hunk 5]
```diff
@@ -186,17 +184,44 @@ FreeProjectWarning = rclass ({name}) ->
...
         quotas = @props.get_total_project_quotas(@props.project_id)
         if not quotas?
             return null
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: This should also be sooner.

File: src/smc-webapp/project_page.cjsx [hunk 6]
```diff
@@ -186,17 +184,44 @@ FreeProjectWarning = rclass ({name}) ->
...
 
+    render_learn_more : ->
+        <a onClick={=>@actions(project_id: @props.project_id).show_extra_free_warning()}> learn more...</a>
+
     render : ->
+        if @props.project
+            status       = JSON.parse(JSON.stringify(@props.project.get('status'))) # without parse and stringify not getting this data in the right format
+            total_quotas = @props.get_total_project_quotas(@props.project_id)
+            memory     = '?'
+            disk       = '?'
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: These should be named better. Are they `memory_used` and `disk_used`?

File: src/smc-webapp/project_page.cjsx [hunk 7]
```diff
@@ -186,7 +184,38 @@ FreeProjectWarning = rclass ({name}) ->
...
+            disk_over_quota = false
+            disk_near_quota = false
+        return [memory_over_quota, memory_near_quota, disk_over_quota, disk_near_quota]
+
     render : ->
+        if not @props.project
+            return null
         if not require('./customize').commercial
             return null
         if @props.free_warning_closed
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: Merge these into a single if.

File: src/smc-webapp/project_page.cjsx [hunk 8]
```diff
@@ -215,8 +245,8 @@ FreeProjectWarning = rclass ({name}) ->
...
-            <Icon name='exclamation-triangle' /> WARNING: This project runs {<span>on a <b>free server (which may be unavailable during peak hours)</b></span> if host} {<span>without <b>internet access</b></span> if internet} &mdash;
-            <a onClick={=>@actions(project_id: @props.project_id).show_extra_free_warning()}> learn more...</a>
+            <Icon name='exclamation-triangle' /> WARNING: {<span>This project is over its memory quota. Either add more memory through upgrades or kill processes.</span> if memory_over_quota} {<span>This project is near its memory quota. Either add more memory through upgrades or kill processes.</span> if memory_near_quota} {<span>This project is over its disk space quota. Either add more disk space through upgrades or delete files.</span> if disk_over_quota} {<span>This project is near its disk space quota. Either add more disk space through upgrades or delete files.</span> if disk_near_quota} {<span>This project runs</span> if host or internet} {<span>on a <b>free server (which may be unavailable during peak hours)</b></span> if host} {<span>without <b>internet access</b></span> if internet} &mdash;
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: Please format this line to be readable.

File: src/smc-webapp/project_page.cjsx [hunk 9]
```diff
@@ -161,21 +161,19 @@ FreeProjectWarning = rclass ({name}) ->
...
     propTypes :
         project_id : rtypes.string
-
-    shouldComponentUpdate : (nextProps) ->
-        return @props.free_warning_extra_shown != nextProps.free_warning_extra_shown or
-            @props.free_warning_closed != nextProps.free_warning_closed or
-            @props.project_map?.get(@props.project_id)?.get('users') != nextProps.project_map?.get(@props.project_id)?.get('users')
+        project    : rtypes.object
 
     extra : (host, internet) ->
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: Naming. Extra what? Does it render?

File: src/smc-webapp/project_page.cjsx [hunk 10]
```diff
@@ -161,21 +161,19 @@ FreeProjectWarning = rclass ({name}) ->
         "#{name}" :
             free_warning_extra_shown : rtypes.bool
             free_warning_closed      : rtypes.bool
+        projects :
+            get_total_project_quotas : rtypes.func
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: You've duplicated this from above.

File: src/smc-webapp/project_page.cjsx [hunk 11]
```diff
@@ -161,21 +161,19 @@ FreeProjectWarning = rclass ({name}) ->
...
 
     propTypes :
         project_id : rtypes.string
-
-    shouldComponentUpdate : (nextProps) ->
-        return @props.free_warning_extra_shown != nextProps.free_warning_extra_shown or
-            @props.free_warning_closed != nextProps.free_warning_closed or
-            @props.project_map?.get(@props.project_id)?.get('users') != nextProps.project_map?.get(@props.project_id)?.get('users')
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: Why did you remove this? It prevents the component from updating when the `project_map` changes arbitrarily.

File: src/smc-webapp/project_page.cjsx [hunk 12]
```diff
@@ -186,7 +184,38 @@ FreeProjectWarning = rclass ({name}) ->
             </ul>
         </div>
 
+    render_learn_more : ->
+        <a onClick={=>@actions(project_id: @props.project_id).show_extra_free_warning()}> learn more...</a>
```
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: Pretty sure you can just pass the function: `@actions(project_id: @props.project_id).show_extra_free_warning`

**Comment by timothyclemansinsea** [comment]
From mobile I removed shouldComponentUpdate because the banner wasn't
showing when one goes directly to the project. I tried various
modifications with no luck

File: src/smc-webapp/project_page.cjsx (hunk 12 above)
Comment (timothyclemansinsea) [inline src/smc-webapp/project_page.cjsx]: Don't know what you want me to do here

File: src/smc-webapp/project_page.cjsx (hunk 8 above)
Comment (timothyclemansinsea) [inline src/smc-webapp/project_page.cjsx]: how would you make this readable? I don't have any good ideas.

**Comment by timothyclemansinsea** [comment]
@johnjeng I made 2 out of 4 requested fixes. See above comments about the other 2.

File: src/smc-webapp/project_page.cjsx (hunk 12 above)
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: Replacing with 
```
onClick={@actions(project_id: @props.project_id).show_extra_free_warning}>
```  
should work I since `show_extra_free_warning` doesn't take any arguments.

File: src/smc-webapp/project_page.cjsx (hunk 8 above)
Comment (johnjeng) [inline src/smc-webapp/project_page.cjsx]: At least break it up into multiple lines.

**Comment by timothyclemansinsea** [comment]
@johnjeng I've addressed everything but the shouldComponentUpdate comment. I don't know how to modify it to show the banner when @props.project becomes defined. I tried various attempts at modifying shouldComponentUpdate with no luck. @williamstein might have some input on this
